### PR TITLE
Minor enhancements to default behavior re: query optimization

### PR DESCRIPTION
**Pull request by SChakravorti21**
Hey! Thank you so much for implementing Substrait support in DuckDB. I've been playing around with it and it's extremely useful.

I ran into some minor hiccups when using the Substrait functions, specifically around how query optimization is handled. This PR contains some enhancements that I hope are not too controversial (happy to make changes based on feedback, though!):

- On the producer side, I disabled the `in_clause` optimization by default. For large `IN` clauses, this optimization rule generates a "mark join" against a `ColumnDataCollection`, and the `ColumnDataCollection` currently cannot be translated to Substrait (it throws this exception: `Internal Error: CHUNK_GET`). In theory, I think we could create a [Virtual Table](https://substrait.io/relations/logical_relations/#virtual-table) inside the Substrait plan, but I'm no database expert and have no clue if this optimization actually makes sense outside of DuckDB.

- I added documentation for the `enable_optimizer` option. Since I wasn't aware of this initially, I tried enabling/disabling query optimization at the connection level, and was surprised to see that it had no effect on the Substrait generation functions. I updated these functions to respect the connection-level settings in case `enable_optimizer` is not explicitly specified. I also added documentation around this new behavior.

I also tried to leave the code cleaner than I found it in some areas:
- Ensured that a clearer error message is provided when parsing a `substrait::Plan` from a JSON string fails.
- Reduced some duplication inside `substrait_extension.cpp`.
- Fixed a typo in the README (`dificulty -> difficulty`).

I ran `make format` and `make test` to adhere to the project standards.

### CODE REVIEW

**Comment by SChakravorti21** [comment]
> In addition, could you also add tests that check the optimizer is properly disabled?

Sure! Should the new tests mainly go under `test/sql/`, or should I also write any tests for the Python/R/etc. APIs specifically?

File: src/substrait_extension.cpp [hunk 1]
```diff
@@ -32,33 +34,45 @@ static void VerifyJSONRoundtrip(unique_ptr<LogicalOperator> &query_plan, Connect
 static void VerifyBlobRoundtrip(unique_ptr<LogicalOperator> &query_plan, Connection &con, ToSubstraitFunctionData &data,
                                 const string &serialized);
 
-static unique_ptr<FunctionData> ToSubstraitBind(ClientContext &context, TableFunctionBindInput &input,
-                                                vector<LogicalType> &return_types, vector<string> &names) {
-	auto result = make_uniq<ToSubstraitFunctionData>();
-	result->query = input.inputs[0].ToString();
-	if (input.named_parameters.size() == 1) {
-		auto loption = StringUtil::Lower(input.named_parameters.begin()->first);
+static bool ShouldGenerateOptimizedSubstrait(const ClientConfig &config,
```
[the remark this answers is not in the document]
Comment (SChakravorti21) [reply src/substrait_extension.cpp:37]: I tried to make it a pure function, so it's technically not _setting_ anything. But I have no objection to renaming it.

**Comment by pdet** [comment]
SQL tests should suffice :-)

**Comment by SChakravorti21** [comment]
I think I've addressed all of the PR comments, and also added several test cases. Please let me know if you think I missed any test cases or if the tests could be improved!

The reason CI was failing was because disabling the `IN` clause rewriter uncovered a minor discrepancy caught by `substrait-validator`. When transforming an `IN` clause:

https://github.com/duckdb/substrait/blob/870bab8725d1123905296bfb1f35ce737434e0b3/src/to_substrait.cpp#L432-L444

The type of the `value` and the types of the `options` can be different. For example, when the expression is of the form `column IN ('lit1', 'lit2')`, where `column` is of DuckDB type `VARCHAR`, `column`'s Substrait type is resolved to `VARCHAR<N>` (`N` depends on the column stats):

https://github.com/duckdb/substrait/blob/870bab8725d1123905296bfb1f35ce737434e0b3/src/to_substrait.cpp#L1002-L1017

Whereas the type of the _literals_ are resolved to Substrait type `STRING`:

https://github.com/duckdb/substrait/blob/870bab8725d1123905296bfb1f35ce737434e0b3/src/to_substrait.cpp#L190-L194

It seems that `substrait-validator` considers these to be separate types and therefore disallows comparing them. This discrepancy probably wasn't caught before because small `IN` clauses are optimized to conjunctions of [`equal`](https://substrait.io/extensions/functions_comparison/#equal), and `equal` is an extension function whose usage could not be validated by `substrait-validator`.

I think the simplest solution is to always use the `STRING` type in Substrait, and this does make all of the existing test cases to pass. I recognize this is not ideal (it loses information gained from the column stats), but would be happy to consider other approaches! I just couldn't think of a better yet generic solution myself.

**Comment by pdet** [comment]
Thanks for all the work! It looks excellent.

Regarding the validator, IIRC we can disable specific warnings and errors per test, maybe that is something to try out?

I would also consider removing it if it starts to become an issue since the project has unfortunately been abandoned and it is not keeping up with the evolving substrait spec.

**Comment by SChakravorti21** [comment]
> Regarding the validator, IIRC we can disable specific warnings and errors per test, maybe that is something to try out?

> I would also consider removing it if it starts to become an issue since the project has unfortunately been abandoned and it is not keeping up with the evolving substrait spec.

These are all valid points! I'm very new to Substrait myself, so I'll leave it to your judgment as to what's the best solution here. The only point I'll add is that DuckDB's `VARCHAR` is semantically equivalent to Substrait's `STRING` type (i.e. strings of "unbounded" size), whereas _Substrait's_ `VARCHAR<N>` is bounded in size. From this point alone, I think the changes in this PR are still relevant, and fortunately we don't have to silence any new warnings or anything. Seeing as there is no alternative to `substrait-validator` at the moment, I do feel it's worth keeping for the time being.

Btw - on an unrelated note, could you shed some light on why the R tests seem to be hanging? I'm not sure if it's related to the changes in this PR. Any tips would be appreciated 🙂

**Comment by SChakravorti21** [comment]
Oh nvm, looks like the R tests have been hanging on `main`, too.

**Comment by pdet** [comment]
Thanks! It looks great.

**Comment by SChakravorti21** [comment]
🎉 

Btw, if I'd like to make additional contributions to this extension, should I start by opening issues? I saw this note on the DuckDB website in regards to this extension, so I was a bit afraid to make an issue in the first place:

> If you have not asked for permission to ask for support, contact us prior to opening an issue. If you open an issue without doing so, we will close it without further review.

**Pull request merged**